### PR TITLE
ProcessRunner to terminate on SIGINT and SIGTERM

### PR DESCRIPTION
**Pull request by dumpstate**
1. Adding SIGINT and SIGTERM OS signal handlers to the main process of a daemon. On a signal, daemon calls `pool.terminate` - that is sending SIGTERM to all processes in the pool.
2. Adding a call to `exit` at the end of the daemon loop - otherwise, the worker processes are not guaranteed to terminate.

### CODE REVIEW

**Review by JP-Globality — APPROVED** [review]

**Review by ronanm-globality — APPROVED** [review]

**Review by german-glob — APPROVED** [review]

File: microcosm_daemon/daemon.py [hunk 1]
```diff
@@ -125,7 +125,11 @@ def start(self):
         """
         self.initialize()
         self.graph.logger.info("Starting daemon {}".format(self.name))
-        self.run_state_machine()
+
+        try:
+            self.run_state_machine()
+        finally:
+            exit(0)
```
Comment (pokey) [inline microcosm_daemon/daemon.py:132]: What if the daemon actually terminated in error? Won't this make it look like everything is fine?

**Review by pokey — COMMENTED** [review]
👋

File: microcosm_daemon/daemon.py (hunk 1 above)
Comment (dumpstate) [reply microcosm_daemon/daemon.py:132]: Hey!!

That's a good point! I have assumed, It should be terminated earlier, with `exit(non-zero)`. But to be fair, I should rather do a lovely try/else, so it's clear from the perspective of the daemon loop:

```
try:
    self.run_state_machine()
except:
    # log err
    exit(1)
else:
    exit(0)
```

Comment (pokey) [reply microcosm_daemon/daemon.py:132]: Yeah that looks good to me 👍😊

File: microcosm_daemon/runner.py [hunk 2]
```diff
@@ -35,14 +36,35 @@ def __init__(self, processes, target, *args, **kwargs):
...
+    def close(self, terminate=False):
+        if self.pool is not None:
+            if terminate:
+                self.pool.terminate()
+            else:
+                self.pool.close()
+
+            try:
+                self.pool.join()
+            except KeyboardInterrupt:
```
Comment (tonybajan-globality) [inline microcosm_daemon/runner.py:64]: is this reachable any more now that you are handing SIGINT explicitly?

Comment (dumpstate) [reply microcosm_daemon/runner.py]: 👍

**Pull request merged**